x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/api/filtering/ApiFilteringActionFilter.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.api.filtering;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.support.ActionFilter;
+import org.elasticsearch.action.support.ActionFilterChain;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationField;
+
+public abstract class ApiFilteringActionFilter<Res extends ActionResponse> implements ActionFilter {
+
+    private final ThreadContext threadContext;
+    private final String actionName;
+    private final Class<Res> responseClass;
+
+    protected ApiFilteringActionFilter(ThreadContext threadContext, String actionName, Class<Res> responseClass) {
+        assert threadContext != null : "threadContext cannot be null";
+        assert actionName != null : "actionName cannot be null";
+        assert responseClass != null : "responseClass cannot be null";
+        this.threadContext = threadContext;
+        this.actionName = actionName;
+        this.responseClass = responseClass;
+    }
+
+    @Override
+    public int order() {
+        return 0;
+    }
+
+    @Override
+    public <Request extends ActionRequest, Response extends ActionResponse> void apply(
+        Task task,
+        String action,
+        Request request,
+        ActionListener<Response> listener,
+        ActionFilterChain<Request, Response> chain
+    ) {
+        final ActionListener<Response> responseFilteringListener;
+        if (isOperator() == false && actionName.equals(action)) {
+            responseFilteringListener = listener.map(this::filter);
+        } else {
+            responseFilteringListener = listener;
+        }
+        chain.proceed(task, action, request, responseFilteringListener);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <Response extends ActionResponse> Response filter(Response response) throws Exception {
+        if (response.getClass().equals(responseClass)) {
+            return (Response) filterResponse((Res) response);
```

Comment:
Are the casts necessary? Seems like the types match.

Reply:
I _think_ that they are necessary. The problem is that ActionFilter doesn't have class-level generic types I could take advantage of -- they're just on the `apply` method. I introduced generic types at the class level of ApiFilteringActionFilter so that I could use them throughout the class. They're the same as the ones for the `apply` method, but the compiler doesn't know that. So I have to do the cast. I think the fix would be to add types at the class level for ActionFilter, but I wasn't sure what all that might break (since it can be used in 3rd-party plugins). Let me know if you're seeing something I'm not seeing though.

Reply:
Ah sorry, I see it is necessary for the argument to cast, but since Res is already a Response, no cast should be necessary?

Reply:
The compiler disagrees with you :)
```
> Task :x-pack:plugin:core:compileJava FAILED
/Users/kmassey/workspace/elasticsearch/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/api/filtering/ApiFilteringActionFilter.java:59: error: incompatible types: Res cannot be converted to Response
            return filterResponse((Res) response);
                                 ^
  where Res,Response are type-variables:
    Res extends ActionResponse declared in class ApiFilteringActionFilter
    Response extends ActionResponse declared in method <Response>filter(Response)
1 error
```

Reply:
Ok, I see, the names are quite confusing. :) Both Res and Response implement ActionFilter, but there is no association between the two. We can improve this in the future.

Reply:
I'm going to go ahead and merge this so that I can get some other PRs going, but let me know if I'm missing something and I'll fix this in a followup PR.